sonar-db/src/main/java/org/sonar/db/version/CreateTableBuilder.java

```diff
@@ -0,0 +1,280 @@
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.db.version;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+import java.util.stream.Stream;
+import javax.annotation.CheckForNull;
+import org.sonar.core.util.stream.Collectors;
+import org.sonar.db.dialect.Dialect;
+import org.sonar.db.dialect.H2;
+import org.sonar.db.dialect.MsSql;
+import org.sonar.db.dialect.MySql;
+import org.sonar.db.dialect.Oracle;
+import org.sonar.db.dialect.PostgreSql;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Stream.of;
+import static org.sonar.db.version.Validations.CONSTRAINT_NAME_MAX_SIZE;
+import static org.sonar.db.version.Validations.TABLE_NAME_MAX_SIZE;
+import static org.sonar.db.version.Validations.checkDbIdentifier;
+
+public class CreateTableBuilder {
+
+  private final Dialect dialect;
+  private final String tableName;
+  private final List<ColumnDef> columnDefs = new ArrayList<>();
+  private final List<ColumnDef> pkColumnDefs = new ArrayList<>(2);
+  private final Multimap<ColumnDef, ColumnFlag> flagsByColumn = HashMultimap.create(1, 1);
+  @CheckForNull
+  private String pkConstraintName;
+
+  public CreateTableBuilder(Dialect dialect, String tableName) {
+    this.dialect = requireNonNull(dialect, "dialect can't be null");
+    this.tableName = checkDbIdentifier(tableName, "Table name", TABLE_NAME_MAX_SIZE);
+  }
+
+  public List<String> build() {
+    checkState(!columnDefs.isEmpty() || !pkColumnDefs.isEmpty(), "at least one column must be specified");
+
+    return Stream.concat(of(createTableStatement()), createOracleAutoIncrementStatements())
+      .collect(Collectors.toList());
+  }
+
+  public CreateTableBuilder addColumn(ColumnDef columnDef) {
+    columnDefs.add(requireNonNull(columnDef, "column def can't be null"));
+    return this;
+  }
+
+  public CreateTableBuilder addPkColumn(ColumnDef columnDef, ColumnFlag... flags) {
```

Comment:
Still, as you can see with the enum `Flag` and the property `flagsByColumn`, I've left open the door for other flag and support of them for columns.